storage/replica.go

```diff
@@ -1421,20 +1421,25 @@ func (r *Replica) handleRaftReady() error {
 	batch := r.store.Engine().NewBatch()
 	defer batch.Close()
 	if !raft.IsEmptySnap(rd.Snapshot) {
+		// TODO(peter): Could use a "distinct" batch within applySnapshot, but we
+		// also need the Batch.Defer mechanism.
 		var err error
 		if lastIndex, err = r.applySnapshot(batch, rd.Snapshot); err != nil {
 			return err
 		}
 		// TODO(bdarnell): update coalesced heartbeat mapping with snapshot info.
 	}
+	// We know that all of the writes from here forward will be to distinct keys.
```

Comment:
Yes, I think so (though of course ideally we also want to use `Distinct()` in applySnapshot). I'm still somewhat uncomfortable with the semantics of `Distinct()` - they essentially require you to know about all prior writes to the batch. Remind me, why didn't we make `Distinct()` read from the underlying batch? It was because that opened up new questions about what would happen if the batch was written to while a Distinct batch was also active?
I'm thinking maybe the best way to go about it would be making the batch inactive as long as it has an open `Distinct()` batch (i.e. a write to the underlying batch would fail), and making `distinctBatch.Commit()` do nothing but flush to the underlying batch (which would then have to be committed)? The upshot would be that `Distinct()` wouldn't have to worry about what happened to the batch earlier, and that use of the original batch is disallowed until we're done with `Distinct()` mode. I think that's congruent with our usage and avoids the weird global perspective that's currently required for reasoning about when `Distinct()` applies.